lib/bap_llvm/llvm_binary.hpp

```diff
@@ -128,54 +130,58 @@ typedef SymbolRef::Type kind_type;
 
 struct symbol {
     std::string name;
-    kind_type kind;
+    bool is_fun;
+    bool is_debug;
     uint64_t addr;
     uint64_t size;
 };
 
-symbol make_symbol(const SymbolRef &sym, uint64_t size) {
-    auto name = name_or_default(sym);
-    auto addr = addr_or_default(sym);
-    auto type = get_type(sym);
-    return symbol{name, type, addr, size};
-}
+typedef std::vector<symbol> symbols;
 
-symbol make_symbol(const SymbolRef &sym, uint64_t size, uint64_t addr) {
-    auto name = name_or_default(sym);
+error_or<symbol> make_symbol(const SymbolRef &sym, uint64_t size, uint64_t addr) {
+    auto name = get_name(sym);
     auto type = get_type(sym);
-    return symbol{name, type, addr, size};
+    if (!name) return copy_message<symbol>(name);
+    if (!type) return copy_message<symbol>(type);
+    bool is_fun = *type == SymbolRef::ST_Function;
+    bool is_dbg = *type == SymbolRef::ST_Debug;
+    return make_value(symbol{*name, is_fun, is_dbg, addr, size});
 }
 
-std::vector<symbol> read(const ObjectFile &obj) {
-    std::vector<symbol> symbols;
-    auto symbol_sizes = getSymbolSizes(obj);
-    for (auto it : symbol_sizes)
-        symbols.push_back(make_symbol(it.first, it.second));
-    return symbols;
+error_or<symbol> make_symbol(const SymbolRef &sym, uint64_t size) {
+    auto addr = get_addr(sym);
+    if (!addr) return copy_message<symbol>(addr);
+    return make_symbol(sym, size, *addr);
 }
 
-std::vector<symbol> read(const COFFObjectFile &obj) {
-    std::vector<symbol> symbols;
+error_or<symbols> read(const ObjectFile &obj) {
     auto symbol_sizes = getSymbolSizes(obj);
-    for (auto it : symbol_sizes) {
-        uint64_t addr = addr_or_default(it.first, obj);
-        symbols.push_back(make_symbol(it.first, it.second, addr));
+    if (!symbol_sizes) return copy_message<symbols>(symbol_sizes);
+    symbols syms;
+    for (auto it : *symbol_sizes) {
+        if (auto sym = make_symbol(it.first, it.second))
+            syms.push_back(*sym);
     }
-    return symbols;
+    return make_value(std::move(syms));
 }
 
-template <typename ELFT>
-std::vector<symbol> read(const ELFObjectFile<ELFT> &obj) {
-    std::vector<symbol> symbols;
+error_or<symbols> read(const COFFObjectFile &obj) {
     auto symbol_sizes = getSymbolSizes(obj);
-    for (auto it : symbol_sizes)
-        symbols.push_back(make_symbol(it.first, it.second));
-    return symbols;
+    if (!symbol_sizes)
+        return copy_message<symbols>(symbol_sizes);
+    symbols syms;
+    for (auto it : *symbol_sizes) {
+        if (auto addr = get_addr(it.first, obj))
+            if (auto sym = make_symbol(it.first, it.second, *addr))
```

Comment:
Loosing diagnostics. Should be:

```c++

auto addr = get_addr(it.first, obj);
if (failure error = addr) warnings.push_back(error << "skipping symbol" << it.first);
```
Here the main problem is with the `warnings`, that is a collection of failures that were ignored. You can either pass it as a separate argument or, as we planned it originally, report it directly to OCaml side via `bap_llvm_diagnostic` function. The latter is dirty but easy, the former is clean but verbose. A usual tradeoff.